### PR TITLE
added camera shake to ob

### PR DESCRIPTION
**Pull request by Cthulhu80**
# About the pull request

Added a camera shake and stun feature based on the user's distance to the OB. This is a copy of an existing branch that I somehow managed to mess up -> https://github.com/cmss13-devs/cmss13/pull/3605

# Explain why it's good for the game

OBs should feel like OBs, and adding this feature just helps make OBs feel like an actual large explosion. 

# Changelog

:cl:
add: added camera shake and stun to OB.
/:cl:

### CODE REVIEW

File: code/modules/cm_marines/orbital_cannon.dm [hunk 1]
```diff
@@ -398,6 +398,51 @@ var/list/ob_type_fuel_requirements
 		return TRUE
 	return FALSE
 
+/obj/structure/ob_ammo/warhead/proc/handle_ob_shake(turf/epicenter)
```
Comment (Zonespace27) [inline code/modules/cm_marines/orbital_cannon.dm:401]: codedoc

File: code/modules/cm_marines/orbital_cannon.dm [hunk 2]
```diff
@@ -398,6 +398,51 @@ var/list/ob_type_fuel_requirements
...
+			frequency = 1
+			max_shake_factor = 8
+			max_knockdown_time = 3
+		if("explosive")
+			frequency = 3
+			max_shake_factor = 15
+			max_knockdown_time = 6
+		if("cluster")
+			frequency = 2
+			max_shake_factor = 1
```
Comment (Zonespace27) [inline code/modules/cm_marines/orbital_cannon.dm:421]: use subtypes correctly

File: code/modules/cm_marines/orbital_cannon.dm [hunk 3]
```diff
@@ -398,6 +398,51 @@ var/list/ob_type_fuel_requirements
...
+
+	for(var/mob/living/M in urange(radius_size, epicenter))
```
Comment (Zonespace27) [inline code/modules/cm_marines/orbital_cannon.dm:423]: no one letter variables

File: code/modules/cm_marines/orbital_cannon.dm [hunk 4]
```diff
@@ -398,6 +398,51 @@ var/list/ob_type_fuel_requirements
...
+		var/total_shake_factor = abs(max_shake_factor * distance_percent)
+
+		// it's of type cluster.
+		if(!max_knockdown_time)
+			shake_camera(M, 0.5, total_shake_factor, frequency)
+			continue
+
+		shake_camera(M, total_shake_factor, frequency)
+
+		var/total_stun_time = max_knockdown_time * distance_percent
```
Comment (Zonespace27) [inline code/modules/cm_marines/orbital_cannon.dm:438]: this doesn't need to be a var

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/modules/admin/tabs/event_tab.dm [hunk 5]
```diff
@@ -897,7 +894,12 @@
 		if(alert("Are you SURE you want to do this? It will create an OB explosion!",, "Yes", "No") != "Yes") return
```
Comment (Drulikar) [inline code/modules/admin/tabs/event_tab.dm:894]: Since the qdel is removed from above, we ought to be sure the warhead is deleted if the admin changes their mind. Make sure it is either qdeleted or has QDEL_IN set for all scenarios.

File: code/modules/cm_marines/orbital_cannon.dm [hunk 6]
```diff
@@ -398,10 +401,40 @@ var/list/ob_type_fuel_requirements
...
+		var/distance_percent = ((radius_size - distance) / radius_size)
+
+		var/total_shake_factor = abs(max_shake_factor * distance_percent)
+
+		// it's of type cluster.
+		if(!max_knockdown_time)
+			shake_camera(user, 0.5, total_shake_factor, frequency)
+			continue
+
+		shake_camera(user, total_shake_factor, frequency)
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:422]: Are your parameters correct here? total_shake_factor is currently positionally set to be steps, and frequency is positionally set to be strength. Use named arguments e.g. `strength = total_shake_factor`.

File: code/modules/cm_marines/orbital_cannon.dm [hunk 7]
```diff
@@ -398,10 +401,40 @@ var/list/ob_type_fuel_requirements
...
+/obj/structure/ob_ammo/warhead/proc/handle_ob_shake(turf/epicenter)
+
+	var/radius_size = 30
+
+	for(var/mob/living/user in urange(radius_size, epicenter))
+
+		var/distance = get_accurate_dist(get_turf(user), epicenter)
+
+		var/distance_percent = ((radius_size - distance) / radius_size)
+
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:414]: Must we have all this whitespace? Just group the variables together. Its just kinda weird that almost everything in this proc is double spaced.

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/modules/cm_marines/orbital_cannon.dm [hunk 8]
```diff
@@ -398,10 +401,37 @@ var/list/ob_type_fuel_requirements
 		return TRUE
 	return FALSE
 
+// proc designed for handling ob camera shakes, takes the target location as input and calculates camera shake based off user location.
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:404]: Codedoc it (use /// not //)

File: code/modules/cm_marines/orbital_cannon.dm [hunk 9]
```diff
@@ -353,6 +353,9 @@ var/list/ob_type_fuel_requirements
 /obj/structure/ob_ammo/warhead
 	name = "theoretical orbital ammo"
 	var/warhead_kind
+	var/frequency
+	var/max_shake_factor
+	var/max_knockdown_time
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:358]: Could codedoc these (/// before each) though ultimately the minimum you should change is the name of `frequency` to something like `shake_frequency`

File: code/modules/cm_marines/orbital_cannon.dm [hunk 10]
```diff
@@ -419,24 +449,32 @@ var/list/ob_type_fuel_requirements
 	var/datum/cause_data/cause_data = create_cause_data(initial(name), source_mob)
 	cell_explosion(target, clear_power, clear_falloff, EXPLOSION_FALLOFF_SHAPE_LINEAR, null, cause_data) //break shit around
 	sleep(clear_delay)
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:451]: It seems a little weird this explosion doesn't cause a shake until after the sleep.

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/modules/cm_marines/orbital_cannon.dm (hunk 10 above)
Comment (Cthulhu80) [reply code/modules/cm_marines/orbital_cannon.dm:451]: the code for the main explosion is right under and it only executes if the turf is valid. This is a much smaller initial explosion and I feel like this shouldn't warrant the camera shake unless the main larger explosion executes.

File: code/modules/cm_marines/orbital_cannon.dm [hunk 11]
```diff
@@ -419,24 +449,32 @@ var/list/ob_type_fuel_requirements
...
 		return
 
+	// Checks turf around the target
 	for(var/turf/T in range(2, target))
 		if(!T.density)
 			cell_explosion(target, standard_power, standard_falloff, EXPLOSION_FALLOFF_SHAPE_LINEAR, null, cause_data)
+			handle_ob_shake(target)
 			sleep(double_explosion_delay)
 			cell_explosion(target, standard_power, standard_falloff, EXPLOSION_FALLOFF_SHAPE_LINEAR, null, cause_data)
-			return
```
Comment (Drulikar) [inline code/modules/cm_marines/orbital_cannon.dm:434]: Why this change? This will mean HE is now multiple explosions and thus a big buff to its power.

File: code/modules/cm_marines/orbital_cannon.dm (hunk 10 above)
Comment (Drulikar) [reply code/modules/cm_marines/orbital_cannon.dm:451]: I suppose, but cell_explosion causes a sfx, and all other instances of cell_explosion here are with a ob_shake. So its weird this is the only one that has a sound, then delay, then there's a shake from a different explosion.

Just experiment with this. If you're happy with it as is, just let me know here and its resolved.

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/modules/cm_marines/orbital_cannon.dm (hunk 10 above)
Comment (Drulikar) [reply code/modules/cm_marines/orbital_cannon.dm:451]: I suppose, but cell_explosion causes a sfx, and all other instances of cell_explosion here are with a ob_shake. So its weird this is the only one that has a sound, then delay, then there's a shake from a different explosion.

Just experiment with this. If you're happy with it as is, just let me know here and its resolved.

Comment (Cthulhu80) [reply code/modules/cm_marines/orbital_cannon.dm:451]: I'm happy with it.

File: code/modules/cm_marines/orbital_cannon.dm (hunk 11 above)
Comment (Cthulhu80) [reply code/modules/cm_marines/orbital_cannon.dm:434]: That would be an accidental bug that I introduced, gonna fix that.

**Review by Drulikar — APPROVED** [review]
Ready for test merge (See Morrow's request here https://github.com/cmss13-devs/cmss13/pull/3605#pullrequestreview-1475161424 )

**Comment by Drulikar** [comment]
Sounds like Morrow is happy with the previous test merge, so its good to go then.
